src/middlewared/middlewared/plugins/pool.py

```diff
@@ -1377,91 +1376,55 @@ async def import_pool(self, job, data):
-                pool = p
-                break
-        if pool is None:
-            raise CallError(f'Pool with guid "{data["guid"]}" not found', errno.ENOENT)
-
+        # import zpool
         try:
-            job.check_pipe("input")
-            key = job.pipes.input.r
-        except ValueError:
-            key = None
-
-        passfile = None
-        if key and data.get('passphrase'):
-            encrypt = 2
-            passfile = tempfile.mktemp(dir='/tmp/')
-            with open(passfile, 'w') as f:
-                os.chmod(passfile, 0o600)
-                f.write(data['passphrase'])
-        elif key:
-            encrypt = 1
-        else:
-            encrypt = 0
-
-        pool_name = data.get('name') or pool['name']
-        pool_id = None
-        try:
-            pool_id = await self.middleware.call('datastore.insert', 'storage.volume', {
-                'vol_name': pool_name,
-                'vol_encrypt': encrypt,
-                'vol_guid': data['guid'],
-                'vol_encryptkey': str(uuid.uuid4()) if encrypt else '',
-            })
-            pool = await self.middleware.call('pool.query', [('id', '=', pool_id)], {'get': True})
-            if encrypt > 0:
-                if not os.path.exists(GELI_KEYPATH):
-                    os.mkdir(GELI_KEYPATH)
-                with open(pool['encryptkey_path'], 'wb') as f:
-                    f.write(key.read())
-
-            await self.middleware.call('pool.scrub.create', {
-                'pool': pool_id,
-            })
-
-            await self.middleware.call('zfs.pool.import_pool', pool['guid'], {
-                'altroot': '/mnt',
-                'cachefile': ZPOOL_CACHE_FILE,
-            }, True, None, pool_name)
-
-            await self.middleware.call('zfs.dataset.update', pool_name, {
-                'properties': {
-                    'aclinherit': {'value': 'passthrough'},
-                },
-            })
+            guid = data['guid']
+            opts = {'altroot': '/mnt', 'cachefile': ZPOOL_CACHE_FILE}
+            any_host = True
+            use_cachefile = None
+            new_name = data.get('name')
+            await self.middleware.call('zfs.pool.import_pool', guid, opts, any_host, use_cachefile, new_name)
+        except Exception as e:
+            if e.errno == errno.ENOENT:
+                raise CallError(f'Pool with guid "{guid}" not found', errno.ENOENT)
+            else:
+                raise CallError(f'Failed to import pool with guid "{guid}"')
 
-            # We would like to reset mountpoints of all the datasets under this pool to ensure that we don't
-            # have any misconfigured mountpoints but we would meanwhile like to ensure that we don't do this
-            # for datasets which are managed by us like ix-applications as that will cause PVC's to not mount
-            # because legacy mountpoint is expected
-            for child in (await self.middleware.call(
```

Comment:
This is not recursively iterating on all child datasets. It just iterates on first level children similar to the changes you introduced

Reply:
No it's not. It's returning all child datasets recursively.

Reply:
Can you please confirm if it is iterating over all children recursively ;) ?

Reply:
I've already confirmed. It is

Reply:
Let me clarify. The returned dict has `children` keys that have the datasets listed recursively.

Reply:
```
{'properties': {}, 'id': 'ha2/parent', 'type': 'FILESYSTEM', 'name': 'ha2/parent', 'pool': 'ha2', 'children': [{'properties': {}, 'id': 'ha2/parent/child', 'type': 'FILESYSTEM', 'name': 'ha2/parent/child', 'pool': 'ha2', 'children': [{'properties': {}, 'id': 'ha2/parent/child/grandchild', 'type': 'FILESYSTEM', 'name': 'ha2/parent/child/grandchild', 'pool': 'ha2', 'children': [{'properties': {}, 'id': 'ha2/parent/child/grandchild/grand-grand-child', 'type': 'FILESYSTEM', 'name': 'ha2/parent/child/grandchild/grand-grand-child', 'pool': 'ha2', 'children': [{'properties': {}, 'id': 'ha2/parent/child/grandchild/grand-grand-child/grand-grand-grand-child', 'type': 'FILESYSTEM', 'name': 'ha2/parent/child/grandchild/grand-grand-child/grand-grand-grand-child', 'pool': 'ha2', 'children': [{'properties': {}, 'id': 'ha2/parent/child/grandchild/grand-grand-child/grand-grand-grand-child/grand-grand-grand-grand-child', 'type': 'FILESYSTEM', 'name': 'ha2/parent/child/grandchild/grand-grand-child/grand-grand-grand-child/grand-grand-grand-grand-child', 'pool': 'ha2', 'children': []}]}]}]}]}]}

Reply:
```
❯ python3
Python 3.9.2 (default, Feb 28 2021, 17:03:44)
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from middlewared.client import Client as cl
>>> d=c.call('zfs.dataset.query', [['id', '=', 'evo']], {"extra": {"retrieve_properties": False, "flat": False}})
>>> len(d)
1
>>> len(d[0]['children'])
35
>>> exit()
❯ zfs list -r evo | wc -l
864
❯ zfs list -d 1 -H evo | wc -l
36
```

Reply:
>We were iterating every single dataset (recursively) for the imported pool and resetting the mountpoint recursively. This made no sense. All we need to do is iterate the top-level datasets of the zpool and reset the mountpoints ?passing the recursive flag. This, potentially, saves a ton of calls to zfs.dataset.inherit

Ah, i was confused because of this ☝🏻

Reply:
I'll update the verbiage in the comment to not cause confusion. Thanks!